cli/src/cli_util.rs

```diff
@@ -692,20 +692,81 @@ impl CommandHelper {
         WorkspaceCommandHelper::new(ui, self, workspace, repo)
     }
 
+    fn create_and_check_out_recovery_commit(
+        &self,
+        ui: &mut Ui,
+        workspace: &mut Workspace,
+    ) -> Result<Arc<ReadonlyRepo>, CommandError> {
+        let operation = self.resolve_operation(ui, workspace.repo_loader())?;
+        let repo = workspace.repo_loader().load_at(&operation)?;
+        let commit_id = repo
+            .view()
+            .get_wc_commit_id(workspace.workspace_id())
+            .ok_or_else(|| CommandError::InternalError("Could not load commit ID".to_string()))?;
+        let commit = repo.store().get_commit(commit_id)?;
+        let tree_id = commit.tree_id();
+
+        let mut tx = repo.start_transaction(&self.settings);
+        let mut_repo = tx.mut_repo();
+        let new_commit = mut_repo
+            .new_commit(&self.settings, vec![commit_id.clone()], tree_id.clone())
+            .write()?;
+        mut_repo.set_wc_commit(workspace.workspace_id().clone(), new_commit.id().clone())?;
+        let ret = Ok(tx.commit("recovery commit"));
+
+        let mut locked_workspace = workspace.start_working_copy_mutation()?;
```

Comment:
Might need to copy `check_working_copy_writable()` here, but I haven't tried.

Reply:
Just checking...do you mean that we should call `start_working_copy_mutation` before we create the new commit, so that if the working copy is not writable, we can print a message to the user without having written a useless commit?

`start_working_copy_mutation` already calls `check_working_copy_writable` (through `unchecked_start_working_copy_mutation`) so I don't think we need to call it here.

Reply:
> Just checking...do you mean that we should call `start_working_copy_mutation` before we create the new commit, so that if the working copy is not writable, we can print a message to the user without having written a useless commit?

No. (Doing that might be better, but isn't a requirement.)

> `start_working_copy_mutation` already calls `check_working_copy_writable` (through `unchecked_start_working_copy_mutation`)

This `workspace.start_working_copy_mutation()` doesn't go through `WorkspaceCommandHelper`. Perhaps, we can construct `WorkspaceCommandHelper` first by `for_loaded_repo()`/`workspace_helper_no_snapshot()`, and use `unchecked_start_working_copy_mutation()`. (The name "unchecked" is a bit odd, but it does some checks.)

Reply:
Ah, sorry for the wrong information (about `start_working_copy_mutation` calling `check_working_copy_writable` . I've implemented your suggestion. Also now we do call `for_loaded_repo()` before creating the commit, so there will be no useless commit if the working copy is not writable.